designs/2024-multithread-linting/README.md

```diff
@@ -0,0 +1,780 @@
+When running in a worker thread, the new method will submit the name of the config file to the controlling thread, either using the worker's own `MessageChannel` or with a dedicated `BroadcastChannel`.
+The warnings will be collected by the controlling thread, which will call its own implementation of `emitEmptyConfigFileWarning`.
+The controlling thread implementation will keep track of warnings that were already emitted using a `Set` to make sure that they are not re-emitted.
+
+#### Other Warnings
+
+Warnings generated by plugins or other tools will not be emitted unfiltered.
+It will be the responsibility of each plugin to avoid duplicated warnings in multithread mode.
+
+### Tests
+
+New scenarios to be tested include cases where ESLint is launched from the CLI with the `--concurrency` flag, and usages of `ESLint#lintFiles()` with the `concurrency` option.
+
+## Documentation
+
+<!--
+    How will this RFC be documented? Does it need a formal announcement
+    on the ESLint blog to explain the motivation?
+-->
+
+The new `concurrency` option should be documented in the [CLI options](https://eslint.org/docs/latest/use/command-line-interface#options) and in the [`ESLint` constructor options](https://eslint.org/docs/latest/integrate/nodejs-api#-new-eslintoptions).
+A blog post to introduce the new feature would be also good.
+
+## Drawbacks
+
+<!--
+    Why should we *not* do this? Consider why adding this into ESLint
+    might not benefit the project or the community. Attempt to think 
+    about any opposing viewpoints that reviewers might bring up. 
+
+    Any change has potential downsides, including increased maintenance
+    burden, incompatibility with other tools, breaking existing user
+    experience, etc. Try to identify as many potential problems with
+    implementing this RFC as possible.
+-->
+
+### Added Complexity
+
+The new feature will introduce additional complexity, especially because of [option modules](#option-modules).
+
+### False Expectations
+
+There is a legitimate concern that multithread mode may in fact make linting slower in some scenarios because of the extra overhead to initialize each thread.
+This is definitely the case as it has been reported for various related tools.
+
+* pgAdmin/eslint-parallel#4
+* pinterest/esprint#126
+* pinterest/esprint#166
+* origin-1/eslint-p#3
+
+Especially plugins that require a longer initialization time, such as typescript-eslint, will be affected by this problem.
+While several solutions have been proposed to mitigate this issue, discussing them would exceed the scope of this proposal.
+To avoid establishing false expectations, it would make sense to include a note regarding the potential performance degradation in the documentation or perhaps to issue a runtime warning.
+The recommendation will be to enable multithread linting only when it performs measurably faster than linting in single-thread mode.
```

Comment:
Noting again from my comments above, a critical piece of this for users on large machines with large projects and using worker_threads may be ensuring that `UV_THREADPOOL_SIZE` is set high enough to avoid contention.